manticore/core/memory.py

```diff
@@ -283,8 +295,50 @@ def __setitem__(self, index, value):
     def __getitem__(self, index):
         index = self._get_offset(index)
         if isinstance(index, slice):
-            return [bytes([i]) for i in self._data[index]]
-        return bytes([self._data[index]])
+            # return [bytes([i]) for i in self._data[index]]
+            return [Operators.CHR(i) for i in self._data[index]]
+        return Operators.CHR(self._data[index])
+        # return bytes([self._data[index]])
+
+
+class ArrayMap(Map):
+    def __init__(self, address, size, perms, index_bits, backing_array=None, name=None, **kwargs):
+        super(ArrayMap, self).__init__(address, size, perms)
+        if name is None:
+            name = 'ArrayMap_{:x}'.format(address)
+        if backing_array is not None:
+            self._array = backing_array
+        else:
+            self._array = expression.ArrayProxy(expression.ArrayVariable(index_bits, index_max=size, value_bits=8, name=name))
+
+    def __reduce__(self):
+        return self.__class__, (self.start, len(self), self._perms, self._array.index_bits, self._array, self._array.name)
+
+    def __setitem__(self, key, value):
+        self._array[key] = value
+
+    def __getitem__(self, key):
+        return self._array[key]
+
+    def split(self, address):
+        if address <= self.start:
+            return None, self
+        if address >= self.end:
```

Comment:
Is this logic right?

Having a memory page of size 0x100 that starts at address A, the last address we can use is A+0x100-1.
So I assume making a split on `self.end` can't split the addresses.

But making a split on start could in theory return `A[0], A[1:]`.

Reply:
Not sure how; the semantics of `split` is that the address is the start of the newly created "rhs" map. So `split(start)` should return (None, original_map). I also don't think split is used very frequently.

Reply:
> address is the start of the newly created "rhs" map
The 'rhs' is the key here. I kinda assumed it shall be 'lhs'.

So it's fine; might be good to add a docstring that its about rhs :P.

Reply:
The logic is that you can only split on an internal address, not a boundary address. Hence the later assertion on line 329.